[GSB] Avoid emitting associated type diagnostics on the protocol decl

hamishknight opened the pull request:
For explicit abstract protocol floating requirement sources, get the source location from the protocol requirement rather than delegating to the parent `RequirementSource`. This allows us to diagnose on an associated type requirement directly rather than on the protocol declaration.

(Also, clean up a few unnecessary `TypeLoc::withoutLoc()`s)

hamishknight committed:
NFC: Remove a few unnecessary `TypeLoc::withoutLoc()`s

hamishknight committed:
[GSB] Avoid emitting associated type diagnostics on the protocol decl

For explicit abstract protocol floating requirement sources, get the source location from the protocol requirement rather than delegating to the parent `RequirementSource`.

jrose-apple commented:
Great idea, but can't RequirementSource provide this information itself? There's a note in RequirementSource::getLoc about why it's not doing it, but I don't understand it myself.

DougGregor commented on test/NameBinding/scope_map_lookup.swift:
```diff
@@ -23,9 +23,10 @@ protocol P1 {
+protocol AProtocol {
   associatedtype e : e
   // expected-error@-1 {{inheritance from non-protocol, non-class type 'Self.e'}}
+  // expected-error@-2 {{type 'Self.e' constrained to non-protocol, non-class type 'Self.e'}}
```
Much better!

DougGregor reviewed (approved):
Nice improvement, thanks

hamishknight committed:
[GSB] Remove getColonLoc() and getEqualLoc() in favour of getSeparatorLoc()

Most callers aren't concerned about the difference between the two (and those that are are already exercising the appropriate assertions with e.g the use of `getFirstType()`).

hamishknight commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -1678,22 +1678,34 @@ const RequirementSource *FloatingRequirementSource::getSource(
 }
 
 SourceLoc FloatingRequirementSource::getLoc() const {
+  auto getSeparatorLoc = [](const RequirementRepr *req) -> SourceLoc {
```
Good idea! In that case, I think we might as well remove `getColonLoc()` and `getEqualLoc()` in favour of `getSeparatorLoc()`, as most callers aren't concerned about the difference (and those that are are already exercising the appropriate assertion with e.g the use of `getFirstType()`).

hamishknight replied:
Ping @DougGregor

slavapestov commented:
@swift-ci Please smoke test

hamishknight commented:
Is this good to go?
